### PR TITLE
feat: Add "deno_net" extension

### PR DESCRIPTION
**Pull request by bartlomieju**
All TCP, UDP, Datagram, TLS and HTTP ops are now available in `deno_net` crate.

Due to complex interactions in JS API between TCP/UDP/Datagram it's not really feasible
to split those ops into more granular extensions.

TODO:
- [x] move JS code to `extensions/net`
- [x] move `Deno.errors.BadResource` and `Deno.errors.Interrupted` class to `deno_core` so it can be used from `deno_net` extension
- [ ] probably rename files in `extensions/net`
- [x] provide declaration files
- [x] clean dependencies in `runtime/Cargo.toml`

Quirks:
- Lots of duplicated (or, copy-pasted) code: IO ops, unstable checker

### CODE REVIEW

**Comment by bartlomieju** [comment]
@ry I'm pretty much done, only thing left is getting type declaration files sorted out. I would appreciated initial review.

File: extensions/net/01_net.js [hunk 1]
```diff
@@ -3,11 +3,27 @@
...
-  const { read, write } = window.__bootstrap.io;
+  const { BadResource } = core;
+
+  async function read(
+    rid,
+    buffer,
+  ) {
+    if (buffer.length === 0) {
+      return 0;
+    }
```
Comment (ry) [inline extensions/net/01_net.js:14]: Seems unnecessary

File: extensions/net/01_net.js [hunk 2]
```diff
@@ -3,11 +3,27 @@
...
+
+    const nread = await core.opAsync("op_net_read_async", rid, buffer);
+
+    return nread === 0 ? null : nread;
+  }
```
Comment (ry) [inline extensions/net/01_net.js:19]: Nit: too much vertical whitespace

File: extensions/net/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# deno_net
+
+This crate implements networking APIs.
```
Comment (ry) [inline extensions/net/README.md:3]: Ideally a bit more detail here. List the ops it provides?

**Review by ry — COMMENTED** [review]
If I tried to use `Deno.read` on stdin, like this `Deno.read(0, buf)` would it still work after this patch?

**Comment by bartlomieju** [comment]
> If I tried to use `Deno.read` on stdin, like this `Deno.read(0, buf)` would it still work after this patch?

Yes

File: extensions/net/01_net.js (hunk 1 above)
Comment (bartlomieju) [reply extensions/net/01_net.js:14]: It's a copy paste from existing io ops

File: extensions/net/01_net.js (hunk 2 above)
Comment (bartlomieju) [reply extensions/net/01_net.js:19]: Ditto

File: extensions/net/README.md (hunk 3 above)
Comment (bartlomieju) [reply extensions/net/README.md:3]: Added a list of ops

**Review by ry — APPROVED** [review]
LGTM - nice work @bartlomieju, this moves us forward

File: cli/build.rs [hunk 4]
```diff
@@ -68,6 +68,9 @@ fn create_compiler_snapshot(
     "deno.broadcast_channel",
     deno_broadcast_channel::get_declaration(),
   );
+  op_crate_libs.insert("deno.net", deno_net::get_declaration());
+  op_crate_libs
+    .insert("deno.net_unstable", deno_net::get_unstable_declaration());
```
Comment (bartlomieju) [inline cli/build.rs:73]: @kitsonk these are the declaration I mentioned on Discord

File: cli/dts/lib.deno.ns.d.ts [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 /// <reference no-default-lib="true" />
 /// <reference lib="esnext" />
+/// <reference lib="deno.net" />
```
Comment (bartlomieju) [inline cli/dts/lib.deno.ns.d.ts:5]: And they are used here...

File: cli/dts/lib.deno.unstable.d.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 /// <reference no-default-lib="true" />
 /// <reference lib="deno.ns" />
+/// <reference lib="deno.net_unstable" />
```
Comment (bartlomieju) [inline cli/dts/lib.deno.unstable.d.ts:5]: ...and here

**Pull request merged**